### PR TITLE
Add `transform.iree.tile_to_workgroups` to handle workgroup count lowering.

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
To use transform dialect approach with the normal codegen flow add a
new operation that lowers the workgroup count computation to what IREE
expects today by default. This allows using transform dialect with the
default compilation flow for the early stages of the IREE compiler,
using the transform dialect based codegeneration only in the backend.

This PR only adds a lit test. e2e tests are blocked by https://github.com/iree-org/iree/issues/10406 (and
other issues as well).

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp [hunk 1]
```diff
@@ -424,5 +430,218 @@ transform_dialect::ForeachThreadToWorkgroupOp::applyToOne(
...
+    if (isConstantIntValue(tileSize.value(), 0)) {
+      workgroupCount.push_back(workload[tileSize.index()]);
+      continue;
+    }
+    AffineExpr s0, s1;
+    bindSymbols(rewriter.getContext(), s0, s1);
+    auto m = AffineMap::get(0, 2, s0.ceilDiv(s1));
+    Value tileSizeVal =
+        getValueOrCreateConstantIndexOp(rewriter, loc, tileSize.value());
+    Value count = rewriter.create<AffineApplyOp>(
```
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp:481]: Please use `OpFoldResult` and `makeComposedFoldedAffineApply` everywhere possible in general.

File: compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp [hunk 2]
```diff
@@ -424,5 +430,218 @@ transform_dialect::ForeachThreadToWorkgroupOp::applyToOne(
...
+
+  IRRewriter rewriter(getContext());
+  if (failed(lowerWorkgroupCountComputingRegion(rewriter, exportOp.value(),
+                                                mixedTileSizes))) {
+    exportOp.value()->emitOpError("failed to lower workgroup count region");
+    return DiagnosedSilenceableFailure(
+        reportUnknownTransformError(exportOp.value()));
+  }
+
+  /// Everything from here is copied over from TileToForeachThreadOp::apply
```
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp:543]: Please factor this functionality out in core and reuse, it is subject to constant evolution, not looking deeper here for now.

File: compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensionsOps.td [hunk 3]
```diff
@@ -153,4 +153,50 @@ def ForeachThreadToWorkgroupOp : Op<Transform_Dialect,
   }];
 }
 
+def TileToWorkgroupsOp :
+    Op<Transform_Dialect, "iree.tile_to_workgroups_op",
+      [AttrSizedOperandSegments,
+       DeclareOpInterfaceMethods<MemoryEffectsOpInterface>,
+       TransformOpInterface]> {
+  let description = [{
+    Wrapper around `structured.tile_to_foreach_thread_op` for use within IREE.
```
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensionsOps.td:162]: "Please see the doc of tile_to_foreach_thread_op in upstream MLIR for a full description of the op semantics".

File: compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensionsOps.td [hunk 4]
```diff
@@ -153,4 +153,50 @@ def ForeachThreadToWorkgroupOp : Op<Transform_Dialect,
...
+
+    In addition to tile and distribute using `scf.foreach_thread`, lowers the
```
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensionsOps.td:164]: Note that this is not the same "tile and distribute" you have been accustomed to, the thread ids here are virtual and are represented by bbArgs.
There is a further separate step of lowering to concrete workgroup/thread ids.

File: tests/transform_dialect/cpu/matmul.mlir [hunk 5]
```diff
@@ -49,15 +49,6 @@ func.func @matmul_static(
 
 // CODEGEN: hal.executable private @matmul_static_dispatch_0 {
 // CODEGEN:   hal.executable.variant public @embedded_elf_x86_64, target = #executable_target_embedded_elf_x86_64_ {
-//
```
Comment (nicolasvasilache) [inline tests/transform_dialect/cpu/matmul.mlir:52]: Thanks for owning the convention and removing this crutch!

**Review by nicolasvasilache — CHANGES_REQUESTED** [review]
Generally looks good, thanks for unblocking this!

Minor comments + let's be sure to avoid copy-pasta of core code from the get go.

File: compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp [hunk 6]
```diff
@@ -424,5 +430,218 @@ transform_dialect::ForeachThreadToWorkgroupOp::applyToOne(
...
+//===---------------------------------------------------------------------===//
+// TileToWorkgroupsOp
+//===---------------------------------------------------------------------===//
+
+/// Lower the ops within the workgroup count region of `exportOp` that
+/// represents the workgroup count calculation, to the actual
+/// computation that returns the number of workgroups. For now
+/// this lowers the `flow.dispatch.workgroup_count_from_dag_root` op
+/// to `ceilDiv(workload, tileSizes)`.
+static LogicalResult lowerWorkgroupCountComputingRegion(
```
[the remark this answers is not in the document]
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp:442]: :) , at the least I guess I own the fall out here. But broadly this is just to be able to use transform dialect for just codegen using what IREE does at Flow by default. I think it is perfectly valid to inject transform dialect specific dispatch region formation at Flow level and use a separate transform dialect script for the associated backend.

File: compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp (hunk 2 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp:543]: I tried to, but the code here is using several methods that live on the op instance. It seemed pretty hard to decouple. If possible, I'd land this as is, and let folks who understand this better try to unify it?

**Comment by MaheshRavishankar** [comment]
Ill fix the failures here and address some comments. I'd really appreciate help to undo the copy-paste. Didnt seem trivial to me.

File: compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp [hunk 7]
```diff
@@ -424,5 +430,218 @@ transform_dialect::ForeachThreadToWorkgroupOp::applyToOne(
...
+        exportOp,
+        "expected region to have a single "
+        "flow.dispatch.workgroup_count_from_dag_root op");
+  }
+  auto workgroupCountOp = *workgroupCountOps.begin();
+  auto workload = workgroupCountOp.getOperands();
+
+  if (tileSizes.size() > workload.size()) {
+    return rewriter.notifyMatchFailure(
+        exportOp, "tile sizes more than the workload captured");
```
[the remark this answers is not in the document]
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp:464]: Fair enough. I expect this is only for the case where there is no "special dispatch region formation" at the Flow level as well.

File: compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp (hunk 2 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp]: Done now.

**Review by nicolasvasilache — APPROVED** [review]
Much better, thanks Mahesh!

**Pull request merged**